### DISCUSSION TITLE
ListBox Container Generation when ItemsSource Element/Item Type is ListBoxItem

### DISCUSSION THREAD

**Discussion by robloo** [AvaloniaUI/Avalonia#12420] [Q&A] [answered]
In WinUI/UWP (probably WPF as well) it was possible to create an ObservableCollection of ListBoxItem in the view model and then bind that to the ItemsSource of the ListBox. ListBox would detect that the collection's element type is ListBoxItem and consider it as it's own container automatically -- no additional ListBoxItem containers would be generated. Avalonia supports this to a degree:

However, any ListBoxItem generated in the view model then bound to a ListBox is very prone to cause crashes (issue 1) if the ListBox itself is re-parented from code-behind. Since ListBoxItem is a control it also can't be used multiple places in the UI. So the work around is to bind to a view model instead of a ListBoxItem and then let the ListBox generate different containers and you can use the same list multiple places in the UI. But now there is no access to the ListBoxItem container itself.

WinUI/UWP allowed this case as well if the DataTemplate's root type was ListBoxItem. In this case again, the ListBox would see that the DataTemplate type is its own container already and not generate new ones. Avalonia doesn't do this though and will created two nested ListBoxItems (issue 2).

So there seems to be no way to have control of the ListBoxItem based on the view model's state AND share the same collection in multiple places of the UI.

The use case in question requires control of ListBoxItems directly because in some cases (based on another view model property) the ListBoxItem must have IsHitTestVisible=False (these represent group headings in the ListBox, think month name or folder name, etc where clickable items appear next in the list under the heading). 

My questions are:

 1. Does Avalonia support this case easily, am I missing something? I would rather not derive a new ListBox control and control item container generation to achieve this.
 2. Is there a simple work-around where the ItemContainerTheme could set IsHitTestVisible=false based on a view model property supplied in the ItemsSource? In the WinUI world there aren't usually any good ways for a "style" to reach into a bound view model in the items source.

And yes, this probably could be worked-around by creating a second, third, etc. collection of ListBoxItems in the view model. I would rather do anything else than duplication though.

**Comment by jp2masa** [comment, marked as answer]
> Is there a simple work-around where the ItemContainerTheme could set IsHitTestVisible=false based on a view model property supplied in the ItemsSource? In the WinUI world there aren't usually any good ways for a "style" to reach into a bound view model in the items source.

I think bindings always worked in styles (something like https://github.com/AvaloniaUI/Avalonia/issues/2710#issue-464275227).

ItemContainerTheme should also support bindings (something like https://github.com/AvaloniaUI/Avalonia/issues/10718#issuecomment-1476564669).

**Reply by robloo** [reply]
I don't know why I was thinking the ItemContainerTheme had a different DataContext scope. I clearly mixed some things up. But you are indeed correct -- this works and is the easy solution I was missing. Thanks so much for pointing that out!
